File: 06_pattern_matching/python/pattern_matching.py
class Dog:
  def __init__(self, name):
    self.name = name

class Cat:
  def __init__(self, age):
    self.age = age






def classify(pet):
  match pet:
    case Cat(age=age):
      print("A cat that's {} years old".format(age))
    case Dog:
      print("A dog named {}".format(pet.name))

def number(x):
    match x:
        case 1: print("one"),
        case 2 | 3: print("two or three"),
        case x if x >=4: print("four or bigger"), # range matching not yet possible
        case _: print("anything"),



if __name__ == "__main__":
    for x in range(0, 6):
        number(x)
    classify(Dog("Fido"))
    classify(Cat(3))

File: 06_pattern_matching/rust/pattern_matching.rs
struct Dog {
    name: String,
}

struct Cat {
    age: u8,
}

enum Animals {
    Dog(Dog),
    Cat(Cat),
}

fn classify(animal: Animals) {
    match animal {
        Animals::Cat(Cat {age}) => println!("A cat that's {} years old", age),
        Animals::Dog(d) => println!("A dog named {}", d.name),
    }
}

fn number(x: i32) {
    match x {
        1 => println!("one"),
        2 | 3 => println!("two or three"),
        4..=i32::MAX => println!("four or bigger"),
        _ => println!("anything"),
    }
}

fn main() {
    for x in 0..6 { number(x)  }
    classify(Animals::Dog(Dog {
        name: "Fido".to_string(),
    }));
    classify(Animals::Cat(Cat { age: 3 }));
}
